Unit test for consistency of model time changes.

grahamgower opened the pull request:
Closes #889.

The test passes, so I'm not sure what #890 is suggesting. Probably I'm missing something here---the mulitple-merger coalescent maths is all way over my head.

Also, I didn't see any info on how the cunit tests should be grouped/ordered, so possibly I put this in a less than ideal location.

jeromekelleher commented on lib/tests/tests.c:
```diff
@@ -3117,6 +3117,84 @@ compute_beta_coalescence_rate_fails(unsigned int num_ancestors, double alpha)
+}
+
+static void
+test_model_time_change_consistency1(double population_size, double t)
```
The convention is to use the ``test_`` prefix for top-level tests (like in unittest), so let's change this to ``check_model_time_consistency``

jeromekelleher commented on lib/tests/tests.c:
```diff
@@ -3117,6 +3117,84 @@ compute_beta_coalescence_rate_fails(unsigned int num_ancestors, double alpha)
+test_model_time_change_consistency(void)
+{
+    double N;
+    for (N=1.0; N<1e9; N+=N) {
```
This is a **lot** of Ns! Something like ``double Ns = {1e-6, 1e-1, 1, 2, 3, 10, 1e4, 1e9}`` would probably cover all the interesting cases I'd say.

jeromekelleher commented on lib/tests/tests.c:
```diff
@@ -3117,6 +3117,84 @@ compute_beta_coalescence_rate_fails(unsigned int num_ancestors, double alpha)
+    int ret;
+    msp_t msp;
+    const unsigned int n = 10;
+    const double alpha = 1.5;
```
We should probably check this for the cross product of a few plausible values of alpha and truncation point, ie.

```c
double alphas[] = {1e-6, 0.1, 1, 1.5, 2};
double truncation_points[] = {1e-6, 0.1, 1, 5};

for (j = 0; j < sizeof(alphas) / sizeof(*alphas); j++) {
    for (k = 0; k < sizeof(...)) {
        ret = msp_set_simulation_model_beta(&msp, population_size, alphas[j], truncation_points[k]);
￼        CU_ASSERT_EQUAL_FATAL(ret, 0);
￼        check_time_change(&msp, t);
    }
}

grahamgower commented:
I added a check that `model_rate_to_generation_rate` and `generation_rate_to_model_rate` are inverses. This fails for each of the dirac and beta test cases.

I also changed `check_time_change()` to be a macro instead of a function, as this more faithfully preserves line numbers for the failed asserts.

jeromekelleher commented on lib/tests/tests.c:
```diff
@@ -3117,6 +3117,94 @@ compute_beta_coalescence_rate_fails(unsigned int num_ancestors, double alpha)
+    ret = msp_set_simulation_model_wf_ped(&msp, population_size);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    check_time_change(&msp, t);
+    ret = msp_set_simulation_model_dirac(&msp, population_size, psi, c);
```
Be nice to do the same thing here with plausible values for psi and c

grahamgower replied:
Happy to do that if you can suggest some numbers.

jeromekelleher replied:
Let's say psi = {1e-6, 0.01, 0.5, 0.9, 1 -1e-6} and c = {0.01, 1, 5}

jeromekelleher commented:
> I added a check that model_rate_to_generation_rate and generation_rate_to_model_rate are inverses. This fails for each of the dirac and beta test cases.

Good, thanks. That's reassuring.

> I also changed check_time_change() to be a macro instead of a function, as this more faithfully preserves line numbers for the failed asserts.

Clever, I like it.

jeromekelleher commented:
Hey @grahamgower, just wondering what the status of this one is? Be nice if we could rebase and get it merged soon.

grahamgower commented:
I'll update the parameter values to be in the accepted ranges and rebase.

grahamgower commented:
For moderate-large `reference_size` values, the following expression is zero (which is problematic because we divide by this scalar when doing time/rate conversions).
```
scalar = exp(log(alpha) - alpha * log(m) - (alpha - 1) * reference_size)
             * acceptance_rate;
```

I checked the equation, and it faithfully reproduces the previous code before my changes in #904. Is it reasonable to have large values for reference_size, or should that ordinarily be closer to 1?

jeromekelleher commented:
I would have thought reference_size of 10^6 would be perfectly reasonable. @JereKoskela, any thoughts here?

JereKoskela commented:
Sorry, I should have noticed this sooner. The expression should be

`scalar = exp(log(alpha) - alpha * log(m) - (alpha - 1) * log(reference_size))
             * acceptance_rate;`

That is, the reference size enters polynomially like `N^(alpha - 1)`, not exponentially. That should take care of any underflows to zero.

The question of default values is not trivial, because the drawback of the Schweinsberg model is that it requires astronomical effective population sizes for predicting reasonable amounts of diversity. Perhaps we should just put in a default of 10^6 and highlight this in the documentation, but practitioners should expect to need numbers closer to 10^20 in many settings.

grahamgower commented:
I made the change to `log(reference_size)`, and this triggered some gsl integration errors in other tests. I'll keep chipping away at it.

grahamgower commented:
The beta coalescent bits are happy now. But there are still tests/tests.c failures because the dirac coalescent time/rate conversion functions are not yet inverses.

jeromekelleher commented:
Great, thanks @grahamgower. How about we skip those dirac tests (comment them out I guess) and register the problem as an issue? I'd like to get these open PRs merged as we want to start using Black/clang-format for formatting the code.

grahamgower commented:
Ok. Squashed and ready for merging once circle CI is happy.

jeromekelleher commented:
Done, thanks @grahamgower. I'll make a note on #186 to say we should enable these tests once it's fixed.
